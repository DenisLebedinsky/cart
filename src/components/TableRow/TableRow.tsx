import * as React from "react";

import css from "./_TableRow.module.scss";

import { TableRowProps } from "./TableRow.type";

class TableRow extends React.Component<TableRowProps> {
  public handleRemoveClick = () => {
    const { id } = this.props;
    this.props.removeItem(id);
  };

  public handleQuantityButtonsClick = (num: number) => () => {
    const { id, quantity } = this.props;
    const newQuantity = quantity + num;

    if (newQuantity < 1) {
      return;
    }

    this.props.updateItem(id, newQuantity);
  };

  public render() {
    const { name, price, quantity } = this.props;
    return (
      <div className={css.tableRow}>
        <div>
          <p>{name}</p>
        </div>
        <div>
          <button onClick={this.handleQuantityButtonsClick(-1)}>
            <p>-</p>
          </button>
          <p>{quantity}</p>
          <button onClick={this.handleQuantityButtonsClick(1)}>
            <p>+</p>
          </button>
        </div>
        <div>
          <p>${price}</p>
        </div>
        <div>
          <button onClick={this.handleRemoveClick}>Remove</button>
        </div>
      </div>
    );
  }
}

export default TableRow;
