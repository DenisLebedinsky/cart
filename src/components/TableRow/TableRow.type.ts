import { removeProduct, updateProduct } from "../../redux/reducers/cart/actions";

export interface TableRowProps {
  id: string;
  name: string;
  quantity: number;
  price: number;
  removeItem: typeof removeProduct;
  updateItem: typeof updateProduct;
}
