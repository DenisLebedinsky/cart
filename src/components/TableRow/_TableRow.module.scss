@import "../../sass/mixins";

.tableRow {
  background-color: #FFFFFF;
  display: flex;

  > div {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 10px;

    > p {
      text-align: center;
      font-size: 18px;
      font-weight: bold;
    }

    &:first-child {
      flex: 4;
    }

    &:nth-child(2) {
      flex: 3;

      > p {
        margin: 0 10px;
        width: 50px;
      }

      > button {
        @include white-button;
        @include round-button;
        display: flex;
        justify-content: center;
        align-items: center;

        > p {
          margin: 0;
        }
      }
    }

    &:nth-child(3) {
      flex: 3;
    }

    &:last-child {
      flex: 1;

      > button {
        @include white-button;
      }
    }
  }
}
