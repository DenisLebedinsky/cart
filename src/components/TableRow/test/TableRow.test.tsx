import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import TableRow from "../TableRow";

describe("TableRow", () => {
  let wrapper: ShallowWrapper<TableRow>;

  const name = "marklar";
  const price = 42;
  const quantity = 13;
  const id = "47";

  let removeItem = jest.fn();
  let updateItem = jest.fn();

  beforeEach(() => {
    removeItem = jest.fn();
    updateItem = jest.fn();

    wrapper = shallow(
      <TableRow
        name={name}
        price={price}
        quantity={quantity}
        id={id}
        removeItem={removeItem}
        updateItem={updateItem}
      />
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call removeItem", () => {
    const instance = wrapper.instance() as TableRow;
    instance.handleRemoveClick();

    expect(removeItem).toBeCalled();
    expect(removeItem).toHaveBeenCalledWith(id);
  });

  it("should not call updateItem when newQuantity is less than 1", () => {
    const instance = wrapper.instance() as TableRow;
    instance.handleQuantityButtonsClick(-13)();

    expect(updateItem).not.toBeCalled();
  });

  it("should call updateItem when newQuantity is greater than 0", () => {
    const instance = wrapper.instance() as TableRow;
    instance.handleQuantityButtonsClick(12)();

    expect(updateItem).toBeCalled();
  });
});
