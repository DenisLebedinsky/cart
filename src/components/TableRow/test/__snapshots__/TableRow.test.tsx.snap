// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should match the snapshot 1`] = `
<div
  className="tableRow"
>
  <div>
    <p>
      marklar
    </p>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      <p>
        -
      </p>
    </button>
    <p>
      13
    </p>
    <button
      onClick={[Function]}
    >
      <p>
        +
      </p>
    </button>
  </div>
  <div>
    <p>
      $
      42
    </p>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
